AwsCryptographyPrimitives/runtimes/python/src/aws_cryptography_primitives/internaldafny/extern/HMAC.py

```diff
@@ -0,0 +1,71 @@
+from cryptography.hazmat.primitives.hmac import HMAC
+from cryptography.hazmat.primitives import hashes
+
+import _dafny
+
+import standard_library.internaldafny.generated.Wrappers as Wrappers
+import aws_cryptography_primitives.internaldafny.generated.Digest
+from aws_cryptography_primitives.internaldafny.generated.AwsCryptographyPrimitivesTypes import (
+    HMacInput
+)
+import aws_cryptography_primitives.internaldafny.generated.HMAC
+from aws_cryptography_primitives.internaldafny.generated.HMAC import *
+
+# Extend generated class
+class default__(aws_cryptography_primitives.internaldafny.generated.HMAC.default__):
+
+  @staticmethod
+  def Digest(input: HMacInput):
+      maybe_hmac = HMac.Build(input.digestAlgorithm)
+      if maybe_hmac.IsFailure():
+        return maybe_hmac.PropagateFailure()
+      hmac = maybe_hmac.Extract()
+      hmac.Init(input.key)
+      hmac.BlockUpdate(input.message)
+      output = hmac.GetResult()
+      return Wrappers.Result_Success(_dafny.Seq(output))
+
+# Extend generated class
+class HMac(aws_cryptography_primitives.internaldafny.generated.HMAC.HMac):
+
+    @staticmethod
+    def Build(digest):
+      output = HMac(digest)
+      return Wrappers.Result_Success(output)
+
+    def __init__(self, digest):
+      if digest.is_SHA__256:
+        self.algorithm = hashes.SHA256()
+      elif digest.is_SHA__384:
+        self.algorithm = hashes.SHA384()
+      elif digest.is_SHA__512:
+        self.algorithm = hashes.SHA512()
+      else:
+        raise ValueError()
+      # self.hmac cannot be used until Init is called
+
+    def Init(self, key):
+      key_bytes = bytes(key)
+      self.hmac = HMAC(key_bytes, self.algorithm)
+      # Store a copy of the HMAC at `Init` state.
+      # The Dafny implementation expects that calling `GetResult` will reset
+      #   the HMAC back to its initial state when `Init` is called.
+      # This is the default behavior for our HMAC libraries in .NET and Java
+      #   but PyCA HMAC does not do this.
+      # We must manually reset the HMAC.
+      self.initial_hmac = self.hmac.copy()
```

Comment:
This looks right, but I'd like to glance at PYCA's docs/code to verify this.

Reply:
PYCA: https://cryptography.io/en/latest/hazmat/primitives/mac/hmac/
Note that the `copy` method docs suggests the result of `copy` is independent of `initial_hmac`

Reply:
Here's a Dafny `ensures` clause on the `GetResult` method that I *think* supports this: https://github.com/aws/aws-cryptographic-material-providers-library/blob/main/AwsCryptographyPrimitives/src/HKDF/HMAC.dfy#L54
I *believe* this is saying "after calling GetResult, the 'input so far' should be empty"